### ISSUE TITLE
Support config files as flags interface

### ISSUE THREAD

**Issue by reactivetype** [guildai/guildai#180]
When using guild `config` dependencies with `yml` config files, the config file defines some default values. These default values are used when no flag overrides them or when the program is not running without guild. 

**Main issue:** the user wants to use config dependency and at the same time expose some of the params via flags. No default value is set in the `guild.yml`'s flags configuration. In this case, `guild compare` shows empty values for the flags. However, the user wants to see them to compare the runs. 

**Example**: in the following example, the guild compare table shows:

```
| x | y | z    | .. metrics |
|    |    | 1.0 | ...   
```

instead, it should show:

```
| x | y | z    | .. metrics |
| 10   | 20   | 1.0 | ...   
```

`guild.yml`

```
requires:
   - config: my_config.yml

flags:
    x: null
    y: null
    z:
       default: 1.0
```

`my_config.yml`

```
x: 10
y: 20
```

**Related issue:**

when user runs `guild run op` with above example, guild may be able to look at the flags if they are present in `config` dependency yml file `my_config.yml` is loaded and show the default value tha are not overriden before prompting to continue. Currently, there is a property for setting a static text value for `null` flag but it may be misleading and error-prone as the user may make mistakes by setting a value that is different from the actual default value in `my_config.yml`.

**Comment by gar1t** [comment]
This is a good observation! I think the problem here is that a `config` dependency wants to be a flags dest, similar to globals (for Python scripts) or command line args. If config files were supported as a proper flags interface, Guild would support importing those flags, as it does with the other interfaces.

Technically speaking, the config resource type doesn't define *flags* - you have to define the flags separately. Once you define the flags and their default values, those values will show up in compare.

I might be inclined to rename this issue something like "Support config files as flags interface"

**Comment by reactivetype** [comment]
In order to use config dependency that is determined at run-time, we can use boolean or enumerate flag to determine which config file to load. To set different default values for the flags depending on which config file was loaded, the user can use `model` namespace for it and override the default flags and reuse common-flags. Here is an example:

```
- config: common-ops
  operations:
    add:
      description: Sum all params from two yml config files
      main: combine
      requires:
        - config: common_config.yml
- config: common-flags
  flags:
    x:
      type: float
      default: 1
    use_config_1:
      type: boolean
      default: false
- model: xyz
  operations:
    $include: common-ops
    add:
      requires:
        - config: config1.yml
      flags:
        $include: common-flags
        use_config_1: true
        a: 5
- model: abc
  operations:
    $include: common-ops
    add:
      requires:
      - config: config2.yml
      flags:
        $include: common-flags
        use_config_1: false
        a: 7
```

**Comment by gar1t** [comment]
This is a good example of how to override the included values, which can be done on an operation-by-operation basis.

You don't need to use models in this case though - this behavior applies at the operation level. So if you had used the default model (empty string) and named your operations e.g. `add-xyz` and `add-abc` you can achieve the same op-specific overrides.

The issue that remains is the ability to import values from these files so that you don't have to maintain them in two locations.

Though I may be missing something here!

**Comment by reactivetype** [comment]
Thanks for clarifying the override scope. It's nice that it is possible to use two ops with same main entry script but different overrides. 

Yes, currently we need to copy default values from yml config file to guild.yml's flags

**Comment by gar1t** [comment]
As soon as 0.7 releases there's a short list of new features to get in and support for config file flag interface is on that list.

**Comment by gar1t** [comment]
Note on implementation: The current support for config-generation via dependencies does NOT use `arg-name` for a flag as the destination in config. So this does not work as you might expect:

``` yaml
op:
  flags:
    foo:
      arg-name: bar.foo
  requires:
    - config: config.yml
```

I would expect Guild to write this file in this case:

``` yaml
bar:
  foo: <val>
```

But it uses the flag name for the target, so this file is written:

``` yaml
foo: <val>
```

This forces you to uses a dotted name for the flag name.

This has to be fixed as a part of this enhancement.

As a side note, we don't record this mapping of flag name to "arg name" for a run but we need to. The current mapping exists only as a "cmd" attr. This might land in a `flags-meta` attr. Remember that runs are always independent of projects.

**Comment by gar1t** [comment]
Update here - this was implemented in 0.7.2. [This example](https://github.com/guildai/guildai/blob/1b7faf03572f5e60d3223a1e29b7d0b3bb037700/examples/flags/guild.yml#L45-L49) shows the syntax.

Closing this for now. If there are any remaining issues, let's reopen.
